pkg/globalnet/controllers/service_controller.go

```diff
@@ -105,6 +105,28 @@ func (c *serviceController) process(from runtime.Object, numRequeues int, op syn
 		return c.onDelete(service)
 	}
 
+	if op == syncer.Update {
+		return c.onUpdate(service)
+	}
+
+	return nil, false
+}
+
+func (c *serviceController) onUpdate(service *corev1.Service) (runtime.Object, bool) {
+	key, _ := cache.MetaNamespaceKeyFunc(service)
+
+	// Log an error if the external-ip of the Globalnet internal service is modified.
+	origService := service.GetLabels()[InternalServiceLabel]
+	if origService != "" {
+		if service.Spec.ExternalIPs[0] != "" {
+			globalIPFromAnnotation := service.GetAnnotations()[GlobalIngressIP]
+			if globalIPFromAnnotation != service.Spec.ExternalIPs[0] {
+				klog.Errorf("ExternalIP %q of Globalnet internal service %q does not match with allocated globalIP %q",
+					service.Spec.ExternalIPs[0], key, globalIPFromAnnotation)
+			}
+		}
```

Comment:
I assume the external IP shouldn't ever be empty, also we should check the array length as a user could have removed it entirely.

This validates the external IP against the global IP stored as an annotation. Ideally we should read the associated `GlobalIngressIP` resource and retrieve it's allocated IP as it's possible the user could change the annotation as well, as nonsensical as that seems.  However I really think we should forego the integrity checking here and do it in subctl diagnose.

```suggestion
		if len(service.Spec.ExternalIPs) == 0 {
		    klog.Errorf("The ExternalIP for Globalnet internal service %q is missing", key)
		} else {
                     globalIPFromAnnotation := service.GetAnnotations()[GlobalIngressIP]
                     if globalIPFromAnnotation != service.Spec.ExternalIPs[0] {
                         klog.Errorf("ExternalIP %q of Globalnet internal service %q does not match with allocated globalIP %q",
                         service.Spec.ExternalIPs[0], key, globalIPFromAnnotation)
                     }
		}
```

Reply:
> I assume the external IP shouldn't ever be empty, also we should check the array length as a user could have removed it entirely.

Ack, done.

> This validates the external IP against the global IP stored as an annotation. Ideally we should read the associated `GlobalIngressIP` resource and retrieve its allocated IP as it's possible the user could change the annotation as well,

I thought about it, but it involves making an API call just to log an error for mismatch. So took this approach.

> as nonsensical as that seems, or change the `GlobalIngressIP`'s allocated IP so they don't match. However I really think we should forego the integrity checking here and do it in subctl diagnose.

As we are just logging the issue and not taking any action, please lets keep this for sometime to get feedback (we can remove it later). Also, I will add support in subctl diagnose in a future PR to flag these issues.

Reply:
But it adds more complexity and untested code that needs to be maintained for little value except it can provide a timestamp if the external IP was changed. But we can do that easily by intercepting change updates and comparing the old and new values and logging if changed.  This code also requires maintaining a label just for this purpose which is unlikely to ever happen in the first place. Plus the check is incomplete b/c it doesn't handle unexpected out-of-band changes to the allocated IP.

I strongly think we should keep such integrity checks in subctl diagnose as we can do a more complete job there and doesn't require scraping logs to find a problem. I also think it's inconsistent to perform this one integrity check in code but not for other cases where integrity can be compromised via user mistake (eg what if the user randomly changes an allocated IP).